File: src/colormanager.cpp
#include "colormanager.h"

#include "confighandlerexception.h"
#include "configparser.h"
#include "logger.h"

#include "libnewsboat-ffi/src/colormanager.rs.h"

using namespace podboat;

namespace newsboat {

ColorManager::ColorManager()
	: rs_object(colormanager::bridged::create())
{
}

void ColorManager::register_commands(ConfigParser& cfgparser)
{
	cfgparser.register_handler("color", *this);
}

void ColorManager::handle_action(std::string_view action,
	const std::vector<std::string>& params)
{
	LOG(Level::DEBUG,
		"ColorManager::handle_action(%s,...) was called",
		action);

	std::vector<rust::Str> rust_params;
	for (const auto& param : params) {
		rust_params.push_back(param);
	}
	auto result = colormanager::bridged::handle_action(*rs_object, rust::Str(action.data(),
				action.size()), rust::Slice<const rust::Str>(rust_params.data(), rust_params.size()));

	switch (result->status) {
	case colormanager::bridged::Status::Ok:
		break;
	case colormanager::bridged::Status::InvalidCommand:
		throw ConfigHandlerException(ActionHandlerStatus::INVALID_COMMAND);
	case colormanager::bridged::Status::TooFewParameters:
		throw ConfigHandlerException(ActionHandlerStatus::TOO_FEW_PARAMS);
	case colormanager::bridged::Status::CustomErrorMessage:
		throw ConfigHandlerException(std::string(result->message));
	}
}

void ColorManager::dump_config(std::vector<std::string>& config_output) const
{
	const auto lines = colormanager::bridged::dump_config(*rs_object);
	for (const auto& line : lines) {
		config_output.push_back(std::string(line));
	}
}

std::map<std::string, std::string> ColorManager::get_stfl_styles() const
{
	const auto rust_styles = colormanager::bridged::get_stfl_styles(*rs_object);
	std::map<std::string, std::string> styles;
	for (const auto& [key, value] : rust_styles) {
		styles.insert_or_assign(std::string(key), std::string(value));
	}
	return styles;
}

} // namespace newsboat

File: rust/libnewsboat-ffi/src/colormanager.rs
use libnewsboat::colormanager::{self, ActionHandlerStatus};

// cxx doesn't allow to share types from other crates, so we have to wrap it
// cf. https://github.com/dtolnay/cxx/issues/496
struct ColorManager(colormanager::ColorManager);

#[cxx::bridge(namespace = "newsboat::colormanager::bridged")]
mod ffi {
    enum Status {
        Ok,
        InvalidCommand,
        TooFewParameters,
        CustomErrorMessage,
    }

    struct ActionHandlerResult {
        status: Status,
        message: String,
    }

    struct KeyValue {
        key: String,
        value: String,
    }

    extern "Rust" {
        type ColorManager;

        fn create() -> Box<ColorManager>;

        fn handle_action(
            colormanager: &mut ColorManager,
            action: &str,
            params: &[&str],
        ) -> Box<ActionHandlerResult>;
        fn dump_config(colormanager: &ColorManager) -> Vec<String>;
        fn get_stfl_styles(colormanager: &ColorManager) -> Vec<KeyValue>;
    }
}

fn create() -> Box<ColorManager> {
    Box::new(ColorManager(colormanager::ColorManager::default()))
}

fn handle_action(
    colormanager: &mut ColorManager,
    action: &str,
    params: &[&str],
) -> Box<ffi::ActionHandlerResult> {
    match colormanager.0.handle_action(action, params) {
        Ok(()) => Box::new(ffi::ActionHandlerResult {
            status: ffi::Status::Ok,
            message: String::new(),
        }),
        Err(ActionHandlerStatus::InvalidCommand) => Box::new(ffi::ActionHandlerResult {
            status: ffi::Status::InvalidCommand,
            message: String::new(),
        }),
        Err(ActionHandlerStatus::TooFewParameters) => Box::new(ffi::ActionHandlerResult {
            status: ffi::Status::TooFewParameters,
            message: String::new(),
        }),
        Err(ActionHandlerStatus::CustomErrorMessage(message)) => {
            Box::new(ffi::ActionHandlerResult {
                status: ffi::Status::CustomErrorMessage,
                message,
            })
        }
    }
}

fn dump_config(colormanager: &ColorManager) -> Vec<String> {
    colormanager.0.dump_config()
}

fn get_stfl_styles(colormanager: &ColorManager) -> Vec<ffi::KeyValue> {
    colormanager
        .0
        .get_stfl_styles()
        .into_iter()
        .map(|(key, value)| ffi::KeyValue { key, value })
        .collect()
}
